Design.md

```diff
@@ -0,0 +1,149 @@
+
+## Communication and auth
+
+Clients communicate via gRPC with a simple protocol: One service with four endpoints. The communication is secured [RusTLS](https://github.com/ctz/rustls) which properly [audited](https://github.com/ctz/rustls/blob/master/audit/TLS-01-report.pdf) and by design supports only modern, secure cipher suites. Only TLS 1.2/1.3 are used, and authentication is done using ECDSA, Ed25519 or RSA. RusTLS will remove support for cipher suites deemed insecure, and simply keeping the version up to date should be sufficient in the future as well. [Tonic](https://github.com/hyperium/tonic) is used to serve gRPC and almost automatically handles protocol buffers, encryption and related concerns.
+
+Authentication is implemented with mTLS. Server and client have different CA roots, which they are expected to exchange in a secure way. (Scenario-wise: the client CA is operated by the same organization that hosts job-runner). The server identifies each client with it's CommonName (CN) field of the certificate. The client CA only issues certificates with CN values to developers that are allowed to access the API. All running jobs are bound to the CN of the client calling `Start`, and only a client with that certificate is allowed to execute operations for that job.
+
+A more complete system should also support certificate revocation lists (CRL), so that individual certificates can be revoked in case they are compromised. Other useful feature would be having more fine-grained access control per certificate, for instance restricting clients to only specific endpoints, executables, and resource limits.
+
+## Endpoints
+
+### Start
+
+Starts a new job by spawning a process from given executable path and arguments. Returns a unique job id (UUID v4), that is used to specify the target job for other endpoints. If the executable is not found or cannot be executed, immediately returns an error.
+
+No security checks are applied to the program and arguments. However, as the job is placed inside an isolated container, it shouldn't be able to do much damage. It can still consume resources in the limits that cgroup-limits allow, and it can flood it's output with thrash data.
+
+### Stop
+
+Cancels a job by sending `SIGKILL`. Stop only returns after the process has been terminated, and returns the status of the job. If job was already terminated before calling stop, the old state is returned without attempting termination.
```

Comment:
> Stop only returns after the process has been terminated

More of a nit, but blocking client on the wire might not be the best UX, I would consider just sending the signal and then letting the client query the status on its own for simplicity.

Reply:
It's actually easier to implement that way as well. I actually wrote the spec like that first time, and then decided that a blocking API is nicer. Buzy-polling a web service seems always a bit ugly to me. I could always implement wait-until-completed API endpoint, but that causes more complexity to the codebase.

Reply:
Reverted back to the non-blocking version.